pkg/liquidity-mining/test/L2BalancerPseudoMinter.test.ts

```diff
@@ -0,0 +1,325 @@
+        await expectEvent.inReceipt(receipt, 'Minted', {
+          recipient: user.address,
+          gauge: gauge.address,
+          minted: mockCheckpointStep,
+        });
+      });
+
+      it('calls gauge user checkpoint', async () => {
+        await expectEvent.inIndirectReceipt(receipt, gauge.interface, 'UserCheckpoint', {
+          user: user.address,
+        });
+      });
+    });
+
+    context('when the amount of tokens to transfer is 0', () => {
+      let receipt: ContractReceipt;
+
+      sharedBeforeEach('ensure there will be tokens to transfer after next checkpoint', async () => {
+        // Accounting on the gauge will not increase after a checkpoint.
+        await gauge.setMockCheckpointStep(0);
+        receipt = await (await pseudoMinter.connect(user).mint(gauge.address)).wait();
+      });
+
+      it('performs no transfers', async () => {
+        expectEvent.notEmitted(receipt, 'Transfer');
+      });
+
+      it('keeps the same minted amount for the user', async () => {
+        const totalBefore = await pseudoMinter.minted(user.address, gauge.address);
+        await pseudoMinter.connect(user).mint(gauge.address);
+        expect(await pseudoMinter.minted(user.address, gauge.address)).to.be.eq(totalBefore);
+      });
+
+      it('skips Minted event', async () => {
+        expectEvent.notEmitted(receipt, 'Minted');
+      });
+
+      it('calls gauge user checkpoint', async () => {
+        await expectEvent.inIndirectReceipt(receipt, gauge.interface, 'UserCheckpoint', {
+          user: user.address,
+        });
+      });
+    });
+  });
+
+  describe('mintMany', () => {
+    const mockCheckpointStep = fp(1);
+    let gauges: Contract[];
+    let gaugeAddresses: string[];
+
+    sharedBeforeEach('setup factory and fund pseudo minter', async () => {
+      await vault.grantPermissionsGlobally([await actionId(pseudoMinter, 'addGaugeFactory')], admin);
+      await pseudoMinter.connect(admin).addGaugeFactory(gaugeFactory.address);
+      await BAL.connect(admin).mint(pseudoMinter.address, fp(100));
+
+      gauges = await Promise.all(
+        range(3).map(() => {
+          return deployGauge(gaugeFactory);
+        })
+      );
+      gaugeAddresses = gauges.map((gauge) => gauge.address);
+    });
+
+    it('reverts with valid gauge from invalid factory', async () => {
+      await gauges[0].setMockFactory(ANY_ADDRESS);
+      await expect(pseudoMinter.connect(user).mintMany(gaugeAddresses)).to.be.revertedWith('INVALID_GAUGE_FACTORY');
+    });
+
+    it('reverts with malicious gauge prentending to come from a valid factory', async () => {
+      const maliciousGauge = await deploy('MockChildChainGauge', { args: ['test'] });
+      await maliciousGauge.setMockFactory(gaugeFactory.address);
+      const gaugesWithMalicious = [...gaugeAddresses, maliciousGauge.address];
+      await expect(pseudoMinter.connect(user).mintMany(gaugesWithMalicious)).to.be.revertedWith('INVALID_GAUGE');
```

Comment:
Try also passing an EOA as the gauge (this will revert with no reason, since the EOA won't return any returndata), and a contract with no `factory` function (also revert with no reason, in this case because the callee reverted).